Include/internal/pycore_object.h

```diff
@@ -280,6 +289,68 @@ extern PyStatus _PyObject_InitState(PyInterpreterState *interp);
 extern void _PyObject_FiniState(PyInterpreterState *interp);
 extern bool _PyRefchain_IsTraced(PyInterpreterState *interp, PyObject *obj);
 
+static inline void
+_Py_INCREF_TYPE(PyTypeObject *type)
+{
+#ifndef Py_GIL_DISABLED
+    Py_INCREF(type);
+#else
+    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
+        assert(_Py_IsImmortal(type));
+        return;
+    }
+
+    _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)_PyThreadState_GET();
+    PyHeapTypeObject *ht = (PyHeapTypeObject *)type;
+
+    // Unsigned comparison so that `ht_id=-1`, which indicates that per-thread
+    // refcounting has been disabled on this type, is handled by the "else".
+    if ((size_t)ht->_ht_id < (size_t)tstate->types.size) {
+#  ifdef Py_REF_DEBUG
+        _Py_INCREF_IncRefTotal();
+#  endif
+        _Py_INCREF_STAT_INC();
+        tstate->types.refcounts[ht->_ht_id]++;
+    }
+    else {
+        // The slow path resizes the thread-local refcount array if necessary.
+        // It handles the ht_id=-1 case to keep the inlinable function smaller.
+        _PyType_IncrefSlow(ht);
+    }
+#endif
+}
+
+static inline void
+_Py_DECREF_TYPE(PyTypeObject *type)
+{
+#ifndef Py_GIL_DISABLED
+    Py_DECREF(type);
+#else
+    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
```

Comment:
```suggestion
    if (_Py_IsImmortal(type)) {
```
Immortal heap types are allowed.

Reply:
Immortal heap types are fine. They either go through the `Py_DECREF()`, which is a no-op, or do some unnecessary work with the per-thread refcounts, which is also fine.

The assumption in the assert is that static types are immortal. We already assert that elsewhere (in `_Py_NewReference`).

We could also write:

```c
if (_Py_IsImmortal(type)) {
    return;
}

assert(_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE));
```

That seems a bit less robust to me in cases where the assumptions are violated, but also fine.

Reply:
Either works. I think they equivalent in terms of checking assumptions.